ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,643 @@
+				proposals: RwLock::new(Default::default()),
+				signer: RwLock::new(Default::default()),
+				machine,
+				step_service: IoService::<Duration>::start()?,
+			});
+
+		if engine.period > 0 {
+			let handler = StepService::new(Arc::downgrade(&engine) as Weak<Engine<_>>,
+			                               Duration::from_secs(our_params.period) / 2 + Duration::from_millis(10));
+			engine.step_service.register_handler(Arc::new(handler)).expect("set_signer: Error registering timeout handler.");
+		}
+
+		return Ok(engine);
+	}
+
+	fn sign_header(&self, header: &Header) -> Result<(Signature, H256), Error> {
+		match self.signer.read().as_ref() {
+			None => {
+				return Err(Box::new("sign_header: No signer available.").into());
+			}
+			Some(signer) => {
+				let digest = header.hash();
+				match signer.sign(digest) {
+					Ok(sig) => {
+						return Ok((sig, digest));
+					}
+					Err(e) => {
+						return Err(Box::new(format!("sign_header: failed to sign header, error: {}", e)).into());
+					}
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		assert_eq!(header.number() % self.epoch_length, 0);
+
+		let mut state = CliqueBlockState::new(
+			match header.number() {
+				0 => NULL_AUTHOR,
+				_ => recover_creator(header)?,
+			},
+			extract_signers(header)?);
+
+		state.calc_next_timestamp(header, self.period);
+
+		Ok(state)
+	}
+
+	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		return self.block_state_by_hash.write().get_mut(hash).cloned()
+	}
+
+	/// get CliqueBlockState for given header, backfill from last checkpoint if needed.
+	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		let mut block_state_by_hash = self.block_state_by_hash.write();
+		if let Some(state) = block_state_by_hash.get_mut(&header.hash()) {
+			return Ok(state.clone());
+		}
+		// If we are looking for an checkpoint block state, we can directly reconstruct it.
+		if header.number() % self.epoch_length == 0 {
+			let state = self.new_checkpoint_state(header)?;
+			block_state_by_hash.insert(header.hash().clone(), state.clone());
+			return Ok(state);
+		}
+		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
+		match self.client.read().as_ref().and_then(|w| { w.upgrade() }) {
+			None => {
+				return Err(From::from("failed to upgrade client reference"));
+			}
+			Some(c) => {
+				let last_checkpoint_number = (header.number() / self.epoch_length as u64) * self.epoch_length;
+				assert_ne!(last_checkpoint_number, header.number());
+
+				let mut chain: &mut VecDeque<Header> = &mut VecDeque::with_capacity((header.number() - last_checkpoint_number + 1) as usize);
+
+				// Put ourselves in.
+				chain.push_front(header.clone());
+
+				// populate chain to last checkpoint
+				let mut last = chain.front().unwrap().clone();
```

Comment:
Shouldn't be unwrapping without a good reason outlined.

Reply:
in line 226 we have just pushed an item to the vecdec in front(), so it is graunteed to exists, 
 I don't see how this can be changed without requiring an rewrite . what's your suggestion?

Reply:
In your case you can do something like this since you know it's safe to unwrap there.

```rs
let mut foo = VecDeque::new();
foo.push_back(42);
foo.front().expect("Just pushed to queue, reference to element must exist; qed");
```
In general you need to "prove" why it is safe to unwrap somewhere.